up-streamer/src/ustreamer.rs

```diff
@@ -0,0 +1,974 @@
+///     number: Some(Number::Ip(vec![192, 168, 1, 100])),
+///     ..Default::default()
+/// };
+/// let local_endpoint = Endpoint::new("local_endpoint", local_authority, local_transport.clone());
+///
+/// // A remote endpoint
+/// let remote_authority = UAuthority {
+///     name: Some("remote".to_string()),
+///     number: Some(Number::Ip(vec![192, 168, 1, 200])),
+///     ..Default::default()
+/// };
+/// let remote_endpoint = Endpoint::new("remote_endpoint", remote_authority, remote_transport.clone());
+///
+/// let streamer = UStreamer::new("hoge", 100);
+///
+/// // Add forwarding rules to endpoint local<->remote
+/// assert_eq!(
+///     streamer
+///         .add_forwarding_rule(local_endpoint.clone(), remote_endpoint.clone())
+///         .await,
+///     Ok(())
+/// );
+/// assert_eq!(
+///     streamer
+///         .add_forwarding_rule(remote_endpoint.clone(), local_endpoint.clone())
+///         .await,
+///     Ok(())
+/// );
+///
+/// // Add forwarding rules to endpoint local<->local, should report an error
+/// assert!(streamer
+///     .add_forwarding_rule(local_endpoint.clone(), local_endpoint.clone())
+///     .await
+///     .is_err());
+///
+/// // Rule already exists so it should report an error
+/// assert!(streamer
+///     .add_forwarding_rule(local_endpoint.clone(), remote_endpoint.clone())
+///     .await
+///     .is_err());
+///
+/// // Try and remove an invalid rule
+/// assert!(streamer
+///     .delete_forwarding_rule(remote_endpoint.clone(), remote_endpoint.clone())
+///     .await
+///     .is_err());
+///
+/// // remove valid routing rules
+/// assert_eq!(
+///     streamer
+///         .delete_forwarding_rule(local_endpoint.clone(), remote_endpoint.clone())
+///         .await,
+///     Ok(())
+/// );
+/// assert_eq!(
+///     streamer
+///         .delete_forwarding_rule(remote_endpoint.clone(), local_endpoint.clone())
+///         .await,
+///     Ok(())
+/// );
+///
+/// // Try and remove a rule that doesn't exist, should report an error
+/// assert!(streamer
+///     .delete_forwarding_rule(local_endpoint.clone(), remote_endpoint.clone())
+///     .await
+///     .is_err());
+/// # }
+/// ```
+pub struct UStreamer {
+    #[allow(dead_code)]
+    name: String,
+    registered_forwarding_rules: ForwardingRules,
+    transport_forwarders: TransportForwarders,
+    transport_forwarder_count: TransportForwarderCount,
+    in_transport_out_authorities: InTransportOutAuthorities,
+    message_queue_size: usize,
+}
+
+impl UStreamer {
+    /// Creates a new UStreamer which can be used to add forwarding rules.
+    ///
+    /// # Parameters
+    ///
+    /// * name - Used to uniquely identify this UStreamer in logs
+    /// * message_queue_size - Determines size of channel used to communicate between `ForwardingListener`
+    ///                        and the worker tasks for each currently endpointd `UTransport`
+    pub fn new(name: &str, message_queue_size: usize) -> Self {
```

Comment:
is there a particular reason why you are using `usize` as the type here? FMPOV the size of the queue should actually be much, much smaller than `usize::MAX`, shouldn't it? I would imagine we are talking in the thousands (u16) at most, or am I mistaken?

Reply:
No real reason other than usize is used to typically represent a size of something.

Is your concern that this number could be made too large when a user is configuring it?

Reply:
My understanding is that `usize` is typically used to make platform specific code for aligning memory addresses and array indices easier to implement. On 64bit architecture, it is a uint64, on a 32bit architecture, it is a uint32.

So, unless there is a compelling reason why the channel size should be aligned with the processor architecture, I guess we should use a type here which represents the requirements from the problem domain. In this case, I guess we should not need to support up to 2^64 messages being buffered by the channel, or do we?

Reply:
> In this case, I guess we should not need to support up to 2^64 messages being buffered by the channel, or do we?

Nope! Definitely not.

Fair point. IMHO `u16` should do the trick

Reply:
I guess the reason you used `usize` instead of `u16` before is the API from [async_std::channel::bounded](https://docs.rs/async-std/latest/async_std/channel/fn.bounded.html)

Reply:
This is true as well!

I suppose I can take a `u16` as input and then cast to `usize` when storing it if the concern is around unintentionally making the channel size too large.

WDYT?

Reply:
I updated this as we discussed to take a u16 instead